### ISSUE TITLE
createExtractorFromFile extract file larger than 2GB

### ISSUE THREAD

**Issue by BTYHunter** [YuJianrong/node-unrar.js#155]
In my case, I would like to extract a 2GB file and use `createExtractorFromFile` instead of `createExtractorFromData`. (from issue #103)
Here is my code.
```javascript
const unrar = require('node-unrar-js')
const extractObject = 'd:/testrar/wj2GB.rar'
const extractor = await unrar.createExtractorFromFile({ filepath: extractObject })
const list = extractor.getFileList() //line 81
const fileHeaders = [...list.fileHeaders] 
```
Error is: 
```
UnrarError: File read error
    at ExtractorFile.getFailException (D:\gitlab\laborer\node_modules\?[4mnode-unrar-js?[24m\dist\js\Extractor.js:187:16)
    at ExtractorFile.openArc (D:\gitlab\laborer\node_modules\?[4mnode-unrar-js?[24m\dist\js\Extractor.js:108:24)
    at ExtractorFile.getFileList (D:\gitlab\laborer\node_modules\?[4mnode-unrar-js?[24m\dist\js\Extractor.js:54:32)
    at extractRAR (D:\gitlab\laborer\example\local\extractLocalObject.js:81:26) {
  reason: ?[32m'ERAR_EREAD'?[39m,
  file: ?[90mundefined?[39m
}
```
when I change wj2GB.rar to a smaller file, the program works well.

**Comment by YuJianrong** [comment]
Interesting, can you edit the file `node_modules/node-unrar-js/dist/js/ExtractorFile.js`, Line 68 - 75:
```Javascript
    read(fd, buf, size) {
        const file = this.fileMap[fd];
        const buffer = Buffer.allocUnsafe(size);
        const readed = fs.readSync(fd, buffer, 0, size, file.pos);
        this.unrar.HEAPU8.set(buffer, buf);
        file.pos += readed;
        return readed;
    }
```
To this code:
```Javascript
    read(fd, buf, size) {
        console.log('read', fd, buf, size);
        try {
        const file = this.fileMap[fd];
        const buffer = Buffer.allocUnsafe(size);
        const readed = fs.readSync(fd, buffer, 0, size, file.pos);
        this.unrar.HEAPU8.set(buffer, buf);
        file.pos += readed;
        } catch (e) {
            console.log('exception on read:', e);
            throw e;
        }
        return readed;
    }
```

Run it again and post the log to this issue?

**Comment by BTYHunter** [comment]
sorry for too late.
I change code as you mention.
Here is the error
```javascript
ReferenceError: readed is not defined
    at ExtractorFile.read (D:\gitlab\laborer\node_modules\node-unrar-js\dist\js\ExtractorFile.js:80:9)
    at _jsRead (D:\gitlab\laborer\node_modules\node-unrar-js\dist\js\unrar.js:9:57985)
    at <anonymous>:wasm-function[233]:0x18834
    at <anonymous>:wasm-function[166]:0x1017b
    at <anonymous>:wasm-function[332]:0x22d90
    at <anonymous>:wasm-function[228]:0x17f7b
    at invoke_iii (D:\gitlab\laborer\node_modules\node-unrar-js\dist\js\unrar.js:9:62942)
    at <anonymous>:wasm-function[398]:0x2694b
    at invoke_viiiii (D:\gitlab\laborer\node_modules\node-unrar-js\dist\js\unrar.js:9:64627)
    at <anonymous>:wasm-function[397]:0x26486
```

**Comment by Jianrong-Yu** [comment]
@BTYHunter Oh, sorry the correct code is:
```Typescript
    read(fd, buf, size) {
        console.log('read', fd, buf, size);
        try {
            const file = this.fileMap[fd];
            const buffer = Buffer.allocUnsafe(size);
            const readed = fs.readSync(fd, buffer, 0, size, file.pos);
            this.unrar.HEAPU8.set(buffer, buf);
            file.pos += readed;
            return readed;
        } catch (e) {
            console.log('exception on read:', e);
            throw e;
        }
    }
```

BTW, what OS and FileSystem used when you run the code? Windows + FAT32? And what version of Nodejs you are using?

**Comment by BTYHunter** [comment]
thanks
OS: windows, NTFS File system.
Node JS Version: v14.15.4

the output is:
```javascript
read 3 5519176 7
read 3 5519183 1
read 3 5795512 7
read 3 5795519 14
...
UnrarError: File read error
    at ExtractorFile.getFailException (D:\gitlab\laborer\node_modules\node-unrar-js\dist\js\Extractor.js:187:16)
    at ExtractorFile.openArc (D:\gitlab\laborer\node_modules\node-unrar-js\dist\js\Extractor.js:108:24)
    at ExtractorFile.getFileList (D:\gitlab\laborer\node_modules\node-unrar-js\dist\js\Extractor.js:54:32)
    at extractRAR (D:\gitlab\laborer\example\local\extractLocalObject.js:73:26) {
  reason: 'ERAR_EREAD',
  file: undefined
}
```

**Comment by DearRabbit** [comment]
Hi, @YuJianrong 

I meet similar problem too. It might be caused by the conversion of int64 in cpp into js number, especially the offset in jsSeek.

For offsetLo > 0x80000000, it appears to be negative in js.
```js
jsSeek: function (handleLo, handleHi, offsetLo, offsetHi, method) {
    return Module.extractor.seek(handleLo + handleHi * 0x100000000, offsetLo + offsetHi * 0x100000000, UTF8ToString(method));
},
```

One solution is to convert int64 offset into uint64 in cpp, i'm not sure if it could work.

You could also use BigInt in conversion, something like
```js
BigInt(lo) & 0xffffffffn + BigInt(hi) * 0x100000000n
```
But I think number is enough for file size(`2^53 – 1`)

**Comment by YuJianrong** [comment]
@DearRabbit Thanks for your detailed description of the bug and fix.
I found it can be just using an unsigned shift operator (`>>>`) to fix it:
``` javascript
(offsetLo >>> 0) + offsetHi * 0x100000000
```
It's fixed and published as `v2.0.1`, please try it to see if it's fixed your issue (I have no large size file for testing).

**Comment by YuJianrong** [comment]
Close this issue since no feedback for a long time.
